File: go/mechanisms/evm/exact/facilitator/permit2_settle_transfer_event_test.go
package facilitator

import (
	"context"
	"errors"
	"fmt"
	"math/big"
	"strings"
	"testing"
	"time"

	"github.com/ethereum/go-ethereum/common"
	goethtypes "github.com/ethereum/go-ethereum/core/types"

	x402 "github.com/x402-foundation/x402/go/v2"
	"github.com/x402-foundation/x402/go/v2/mechanisms/evm"
	"github.com/x402-foundation/x402/go/v2/types"
)

const (
	permit2TestPayer   = "0xa0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0"
	permit2TestPayTo   = "0xb1b1b1b1b1b1b1b1b1b1b1b1b1b1b1b1b1b1b1b1"
	permit2TestToken   = "0x036cbd53842c5426634e7929541ec2318f3dcf7e"
	permit2TestNetwork = "eip155:84532"
	permit2TestAmount  = "1000"
	permit2DummySig    = "0x" +
		"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" +
		"bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb" +
		"1b"
)

// permit2SettleMockSigner is a minimal FacilitatorEvmSigner for Permit2 settle transfer-event tests.
// The payer is treated as a deployed contract so a dummy signature falls through to settle when
// simulation is disabled (same pattern as the upto Permit2 settle suite).
type permit2SettleMockSigner struct {
	receipt *evm.TransactionReceipt
	txHash  string
}

func (m *permit2SettleMockSigner) GetAddresses() []string {
	return []string{"0xf1f1f1f1f1f1f1f1f1f1f1f1f1f1f1f1f1f1f1f1"}
}

func (m *permit2SettleMockSigner) ReadContract(ctx context.Context, address string, abi []byte, functionName string, args ...interface{}) (interface{}, error) {
	return nil, nil
}

func (m *permit2SettleMockSigner) VerifyTypedData(ctx context.Context, address string, domain evm.TypedDataDomain, types map[string][]evm.TypedDataField, primaryType string, message map[string]interface{}, signature []byte) (bool, error) {
	return false, nil
}

func (m *permit2SettleMockSigner) WriteContract(ctx context.Context, address string, abi []byte, functionName string, dataSuffix []byte, args ...interface{}) (string, error) {
	if m.txHash != "" {
		return m.txHash, nil
	}
	return "0x" + strings.Repeat("ab", 32), nil
}

func (m *permit2SettleMockSigner) SendTransaction(ctx context.Context, to string, data []byte) (string, error) {
	return m.WriteContract(ctx, to, nil, "", nil)
}

func (m *permit2SettleMockSigner) WaitForTransactionReceipt(ctx context.Context, txHash string) (*evm.TransactionReceipt, error) {
	if m.receipt != nil {
		return m.receipt, nil
	}
	return &evm.TransactionReceipt{Status: evm.TxStatusSuccess, TxHash: txHash}, nil
}

func (m *permit2SettleMockSigner) GetBalance(ctx context.Context, address string, tokenAddress string) (*big.Int, error) {
	return big.NewInt(999_000_000), nil
}

func (m *permit2SettleMockSigner) GetChainID(ctx context.Context) (*big.Int, error) {
	return big.NewInt(84532), nil
}

func (m *permit2SettleMockSigner) GetCode(ctx context.Context, address string) ([]byte, error) {
	normalized := strings.ToLower(address)
	if normalized == strings.ToLower(permit2TestToken) || normalized == strings.ToLower(permit2TestPayer) {
		return []byte{0x60, 0x60}, nil
	}
	return []byte{}, nil
}

func buildExactPermit2SettleFixture() (types.PaymentPayload, types.PaymentRequirements, *evm.ExactPermit2Payload) {
	now := time.Now().Unix()
	permit2Payload := &evm.ExactPermit2Payload{
		Signature: permit2DummySig,
		Permit2Authorization: evm.Permit2Authorization{
			From: permit2TestPayer,
			Permitted: evm.Permit2TokenPermissions{
				Token:  permit2TestToken,
				Amount: permit2TestAmount,
			},
			Spender:  evm.X402ExactPermit2ProxyAddress,
			Nonce:    "12345",
			Deadline: fmt.Sprintf("%d", now+300),
			Witness: evm.Permit2Witness{
				To:         permit2TestPayTo,
				ValidAfter: fmt.Sprintf("%d", now-600),
			},
		},
	}
	requirements := types.PaymentRequirements{
		Scheme:  evm.SchemeExact,
		Network: permit2TestNetwork,
		Asset:   permit2TestToken,
		Amount:  permit2TestAmount,
		PayTo:   permit2TestPayTo,
	}
	payload := types.PaymentPayload{
		X402Version: 2,
		Accepted:    requirements,
		Payload:     permit2Payload.ToMap(),
	}
	return payload, requirements, permit2Payload
}

func TestSettlePermit2_RejectsUnderpayingTransferEvent(t *testing.T) {
	payload, requirements, permit2Payload := buildExactPermit2SettleFixture()
	txHash := "0x" + strings.Repeat("ab", 32)
	token := common.HexToAddress(permit2TestToken)
	payer := common.HexToAddress(permit2TestPayer)
	payTo := common.HexToAddress(permit2TestPayTo)

	signer := &permit2SettleMockSigner{
		txHash: txHash,
		receipt: &evm.TransactionReceipt{
			Status: evm.TxStatusSuccess,
			TxHash: txHash,
			Logs: []*goethtypes.Log{
				makeTransferLog(token, payer, payTo, big.NewInt(900)),
			},
		},
	}

	_, err := SettlePermit2(context.Background(), signer, payload, requirements, permit2Payload, nil, &Permit2FacilitatorConfig{SimulateInSettle: false})
	assertPermit2SettleTransferMismatch(t, err, txHash)
}

func TestSettlePermit2_RejectsEmptyLogs(t *testing.T) {
	payload, requirements, permit2Payload := buildExactPermit2SettleFixture()
	txHash := "0x" + strings.Repeat("cd", 32)

	signer := &permit2SettleMockSigner{
		txHash: txHash,
		receipt: &evm.TransactionReceipt{
			Status: evm.TxStatusSuccess,
			TxHash: txHash,
			Logs:   []*goethtypes.Log{},
		},
	}

	_, err := SettlePermit2(context.Background(), signer, payload, requirements, permit2Payload, nil, &Permit2FacilitatorConfig{SimulateInSettle: false})
	assertPermit2SettleTransferMismatch(t, err, txHash)
}

func TestSettlePermit2_AcceptsMatchingTransferEvent(t *testing.T) {
	payload, requirements, permit2Payload := buildExactPermit2SettleFixture()
	txHash := "0x" + strings.Repeat("ef", 32)
	token := common.HexToAddress(permit2TestToken)
	payer := common.HexToAddress(permit2TestPayer)
	payTo := common.HexToAddress(permit2TestPayTo)

	signer := &permit2SettleMockSigner{
		txHash: txHash,
		receipt: &evm.TransactionReceipt{
			Status: evm.TxStatusSuccess,
			TxHash: txHash,
			Logs: []*goethtypes.Log{
				makeTransferLog(token, payer, payTo, big.NewInt(1000)),
			},
		},
	}

	resp, err := SettlePermit2(context.Background(), signer, payload, requirements, permit2Payload, nil, &Permit2FacilitatorConfig{SimulateInSettle: false})
	if err != nil {
		t.Fatalf("expected settle success, got error: %v", err)
	}
	if !resp.Success {
		t.Fatalf("expected resp.Success = true, got %+v", resp)
	}
	if resp.Transaction != txHash {
		t.Fatalf("expected transaction %q, got %q", txHash, resp.Transaction)
	}
}

func assertPermit2SettleTransferMismatch(t *testing.T, err error, txHash string) {
	t.Helper()
	if err == nil {
		t.Fatal("expected transfer event mismatch error, got nil")
	}
	se := &x402.SettleError{}
	if !errors.As(err, &se) {
		t.Fatalf("expected *x402.SettleError, got %T: %v", err, err)
	}
	if se.ErrorReason != ErrTransferEventMismatch {
		t.Fatalf("expected reason %q, got %q", ErrTransferEventMismatch, se.ErrorReason)
	}
	if se.Transaction != txHash {
		t.Fatalf("expected transaction %q, got %q", txHash, se.Transaction)
	}
}

File: typescript/packages/mechanisms/evm/test/unit/exact/permit2-settle-transfer-event.test.ts
/**
 * Regression tests: the Permit2 exact settle path must verify the ERC-20
 * Transfer event in the settlement receipt, not just receipt.status.
 *
 * The proxy's settle() does not revert when a non-conforming token underpays
 * (fee-on-transfer) or moves nothing at all, so status alone cannot prove the
 * recipient was paid in full. This mirrors the EIP-3009 settle path, which
 * rejects the same receipts via verifyEip3009TransferEvent.
 */
import { describe, it, expect } from "vitest";
import { encodeAbiParameters, keccak256, padHex, toBytes, toHex, getAddress } from "viem";
import { privateKeyToAccount } from "viem/accounts";
import type { TransactionReceipt } from "viem";
import { settlePermit2 } from "../../../src/exact/facilitator/permit2";
import {
  ErrInvalidTransactionState,
  ErrTransferEventMismatch,
} from "../../../src/exact/facilitator/errors";
import {
  PERMIT2_ADDRESS,
  permit2WitnessTypes,
  x402ExactPermit2ProxyAddress,
} from "../../../src/constants";
import type { ExactPermit2Payload } from "../../../src/types";
import type { FacilitatorEvmSigner } from "../../../src/signer";
import type { PaymentPayload, PaymentRequirements } from "@x402/core/types";

const TRANSFER_TOPIC = keccak256(toBytes("Transfer(address,address,uint256)"));
const MOCK_TX = padHex("0xabc", { size: 32 });

// Real keypair so the EIP-712 signature is genuinely valid (ecrecover path).
const payerAccount = privateKeyToAccount(
  "0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d",
);
const PAYER = payerAccount.address;
const RECEIVER = getAddress("0x2222222222222222222222222222222222222222");
const TOKEN = getAddress("0xE7C3D8C9a439feDe00D2600032D5dB0Be71C3c29");
const NETWORK = "eip155:8453" as const;
const AMOUNT = "1000";

function makeTransferLog(opts: {
  address: `0x${string}`;
  from: `0x${string}`;
  to: `0x${string}`;
  value: bigint;
}) {
  return {
    address: opts.address,
    topics: [TRANSFER_TOPIC, padHex(opts.from, { size: 32 }), padHex(opts.to, { size: 32 })] as [
      `0x${string}`,
      `0x${string}`,
      `0x${string}`,
    ],
    data: encodeAbiParameters([{ type: "uint256" }], [opts.value]),
    blockHash: padHex("0x1", { size: 32 }),
    blockNumber: 1n,
    transactionHash: MOCK_TX,
    transactionIndex: 0,
    logIndex: 0,
    removed: false,
  };
}

function makeReceipt(
  logs: ReturnType<typeof makeTransferLog>[],
  status: "success" | "reverted" = "success",
): TransactionReceipt {
  return {
    status,
    logs,
    blockHash: padHex("0x1", { size: 32 }),
    blockNumber: 1n,
    transactionHash: MOCK_TX,
    transactionIndex: 0,
    cumulativeGasUsed: 0n,
    gasUsed: 0n,
    contractAddress: null,
    from: PAYER,
    to: x402ExactPermit2ProxyAddress,
    logsBloom: toHex(new Uint8Array(256)),
    type: "eip1559",
    effectiveGasPrice: 0n,
  } as unknown as TransactionReceipt;
}

async function makeSignedPayload(): Promise<ExactPermit2Payload> {
  const now = Math.floor(Date.now() / 1000);
  const authorization = {
    from: PAYER,
    permitted: { token: TOKEN, amount: AMOUNT },
    spender: x402ExactPermit2ProxyAddress as `0x${string}`,
    nonce: "123456789",
    deadline: String(now + 3600),
    witness: { to: RECEIVER, validAfter: "0" },
  };

  const signature = await payerAccount.signTypedData({
    domain: { name: "Permit2", chainId: 8453, verifyingContract: PERMIT2_ADDRESS },
    types: permit2WitnessTypes,
    primaryType: "PermitWitnessTransferFrom",
    message: {
      permitted: { token: TOKEN, amount: BigInt(AMOUNT) },
      spender: x402ExactPermit2ProxyAddress,
      nonce: BigInt(authorization.nonce),
      deadline: BigInt(authorization.deadline),
      witness: { to: RECEIVER, validAfter: 0n },
    },
  });

  return { signature, permit2Authorization: authorization };
}

function makeMockSigner(receipt: TransactionReceipt): FacilitatorEvmSigner {
  return {
    getCode: async ({ address }: { address: string }) =>
      getAddress(address) === TOKEN ? "0x6080604052" : "0x",
    writeContract: async () => MOCK_TX,
    waitForTransactionReceipt: async () => receipt,
    getAddresses: () => [],
    readContract: async () => {
      throw new Error("unexpected readContract in settle test (simulation disabled)");
    },
  } as unknown as FacilitatorEvmSigner;
}

function makeRequirements(): PaymentRequirements {
  return {
    scheme: "exact",
    network: NETWORK,
    asset: TOKEN,
    amount: AMOUNT,
    payTo: RECEIVER,
    maxTimeoutSeconds: 3600,
    extra: {},
  } as unknown as PaymentRequirements;
}

function makePaymentPayload(requirements: PaymentRequirements): PaymentPayload {
  return {
    x402Version: 2,
    accepted: requirements,
    payload: undefined,
  } as unknown as PaymentPayload;
}

async function settleWithReceipt(receipt: TransactionReceipt) {
  const permit2Payload = await makeSignedPayload();
  const requirements = makeRequirements();
  const payload = makePaymentPayload(requirements);
  payload.payload = permit2Payload as unknown as Record<string, unknown>;
  return settlePermit2(makeMockSigner(receipt), payload, requirements, permit2Payload);
}

describe("settlePermit2 Transfer event verification", () => {
  it("rejects a receipt whose Transfer delivers less than the permitted amount (fee-on-transfer)", async () => {
    const receipt = makeReceipt([
      makeTransferLog({ address: TOKEN, from: PAYER, to: RECEIVER, value: 900n }),
    ]);

    const result = await settleWithReceipt(receipt);

    expect(result.success).toBe(false);
    expect(result.errorReason).toBe(ErrTransferEventMismatch);
    expect(result.transaction).toBe(MOCK_TX);
  });

  it("rejects a receipt with no Transfer event at all (silent no-op token)", async () => {
    const result = await settleWithReceipt(makeReceipt([]));

    expect(result.success).toBe(false);
    expect(result.errorReason).toBe(ErrTransferEventMismatch);
    expect(result.transaction).toBe(MOCK_TX);
  });

  it("rejects a receipt whose only Transfer moves the full amount to a different recipient", async () => {
    const receipt = makeReceipt([
      makeTransferLog({
        address: TOKEN,
        from: PAYER,
        to: getAddress("0x3333333333333333333333333333333333333333"),
        value: 1000n,
      }),
    ]);

    const result = await settleWithReceipt(receipt);

    expect(result.success).toBe(false);
    expect(result.errorReason).toBe(ErrTransferEventMismatch);
  });

  it("accepts a receipt with the full-payment Transfer event", async () => {
    const receipt = makeReceipt([
      makeTransferLog({ address: TOKEN, from: PAYER, to: RECEIVER, value: 1000n }),
    ]);

    const result = await settleWithReceipt(receipt);

    expect(result.success).toBe(true);
    expect(result.transaction).toBe(MOCK_TX);
    expect(result.payer).toBe(PAYER);
  });

  it("still reports invalid_transaction_state for a reverted receipt", async () => {
    const result = await settleWithReceipt(makeReceipt([], "reverted"));

    expect(result.success).toBe(false);
    expect(result.errorReason).toBe(ErrInvalidTransactionState);
  });
});
